src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,282 @@
+            bool hasErrors;
+
+            if (!best.HasValue)
+            {
+                resultOperatorKind = kind;
+                convertedLeftType = leftType ?? CreateErrorType();
+                convertedRightType = rightType ?? CreateErrorType();
+                returnType = CreateErrorType();
+                hasErrors = true;
+            }
+            else
+            {
+                bool leftNull = left.IsLiteralNull();
+                bool rightNull = right.IsLiteralNull();
+
+                var signature = best.Signature;
+
+                bool isObjectEquality = signature.Kind == BinaryOperatorKind.ObjectEqual || signature.Kind == BinaryOperatorKind.ObjectNotEqual;
+
+                resultOperatorKind = signature.Kind;
+                resultMethod = signature.Method;
+                convertedLeftType = signature.LeftType;
+                convertedRightType = signature.RightType;
+                returnType = signature.ReturnType; // PROTOTYPE(tuple-equality) Do we need a special case for nullable equality? (see BindSimpleBinaryOperator)
+
+                HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                hasErrors = isObjectEquality && !BuiltInOperators.IsValidObjectEquality(Conversions, leftType, leftNull, rightType, rightNull, ref useSiteDiagnostics);
+                diagnostics.Add(node, useSiteDiagnostics);
+            }
+
+            if (hasErrors)
+            {
+                ReportBinaryOperatorError(node, diagnostics, node.OperatorToken, left, right, resultKind);
+            }
+
+            PrepareBoolConversionAndTruthOperator(returnType, node, kind, diagnostics, out Conversion boolConversion, out UnaryOperatorSignature boolOperator);
+            return new TupleBinaryOperatorInfo.Single(convertedLeftType, convertedRightType, resultOperatorKind, resultMethod, boolConversion, boolOperator);
+        }
+
+        /// <summary>
+        /// If a element-wise binary operator returns a non-bool type, we will either:
+        /// - prepare a conversion to bool if one exists
+        /// - prepare a truth operator: op_false in the case of an equality (`a == b` will be lowered to `!((a == b).op_false)) or op_true in the case of inequality
+        /// </summary>
+        private void PrepareBoolConversionAndTruthOperator(TypeSymbol type, BinaryExpressionSyntax node, BinaryOperatorKind binaryOperator, DiagnosticBag diagnostics,
+            out Conversion boolConversion, out UnaryOperatorSignature boolOperator)
+        {
+            BoundExpression comparisonResult = new BoundTupleOperandPlaceholder(node, type);
+            var boolean = GetSpecialType(SpecialType.System_Boolean, diagnostics, node);
+
+            // Is the operand implicitly convertible to bool?
+
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            var conversion = this.Conversions.ClassifyConversionFromExpression(comparisonResult, boolean, ref useSiteDiagnostics);
```

Comment:
>ClassifyConversionFromExpression [](start = 46, length = 32)

Why do we classify conversion from expression when we clearly converting from one type to another here?

Reply:
Also, it feels like it should be enough to classify implicit conversion.

---
In reply to: [170493667](https://github.com/dotnet/roslyn/pull/23356#discussion_r170493667) [](ancestors = 170493667)

Reply:
We're following the same rules as used in `if`. See `BindBooleanExpression`.

---
In reply to: [170493739](https://github.com/dotnet/roslyn/pull/23356#discussion_r170493739) [](ancestors = 170493739,170493667)

Reply:
> We're following the same rules as used in if. See BindBooleanExpression.

I do not see much similarity here. ```if``` deals with expressions, we are not. 

---
In reply to: [170755422](https://github.com/dotnet/roslyn/pull/23356#discussion_r170755422) [](ancestors = 170755422,170493739,170493667)